provider/uprops/src/casemapping.rs

```diff
@@ -0,0 +1,271 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::uprops_serde;
+use icu_casemapping::provider::{CaseMappingV1, CaseMappingV1Marker};
+use icu_casemapping::CaseMappingInternals;
+use icu_codepointtrie::CodePointTrieHeader;
+use icu_provider::prelude::*;
+
+use std::convert::TryFrom;
+use std::fs;
+use std::path::PathBuf;
+
+pub struct CaseMappingDataProvider {
+    case_mapping: CaseMappingInternals<'static>,
+}
+
+/// A data provider reading from .toml files produced by the ICU4C icuwriteuprops tool.
+impl CaseMappingDataProvider {
+    pub fn try_new(path: PathBuf) -> Result<Self, DataError> {
+        let toml_str = fs::read_to_string(&path).unwrap();
+        let toml: uprops_serde::case::Main = toml::from_str(&toml_str).unwrap();
+
+        let trie_data = &toml.ucase.code_point_trie;
+        let trie_header = CodePointTrieHeader::try_from(trie_data)?;
+        let trie_index = &trie_data.index;
+        let trie_data = &trie_data.data_16.as_ref().ok_or_else(|| {
+            DataError::custom("Did not find 16-bit data array for case mapping in TOML")
+        })?;
+        let exceptions = &toml.ucase.exceptions.exceptions;
+        let unfold = &toml.ucase.unfold.unfold;
+
+        let case_mapping = CaseMappingInternals::try_from_icu(
+            trie_header,
+            trie_index,
+            trie_data,
+            exceptions,
+            unfold,
+        )
+        .map_err(|e| {
+            DataError::custom("Could not create CaseMappingInternals").with_display_context(&e)
+        })?;
+
+        Ok(Self { case_mapping })
+    }
+}
+
+impl ResourceProvider<CaseMappingV1Marker> for CaseMappingDataProvider {
+    fn load_resource(
+        &self,
+        _req: &DataRequest,
+    ) -> Result<DataResponse<CaseMappingV1Marker>, DataError> {
+        Ok(DataResponse {
+            metadata: DataResponseMetadata::default(),
+            payload: Some(DataPayload::from_owned(CaseMappingV1 {
+                casemap: self.case_mapping.clone(),
+            })),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
suggestion: include a few more more tests

The tests `StringCaseTest::TestGreekUpper` and `StringCaseTest::TestArmenian()` seem like short tests to implement. The latter depends on whether we match ICU on Eastern vs. Western Armenian.

Also, `StringCaseTest::TestLongUpper()` seems interesting and short. Maybe Rust makes that a non-issue, but nice to have.

All the other tests seem like they are about titlecasing, UTF-8 and/or Edits, so no other tests of interest.

Reply:
TestArmenian depends on enabling the Armenian code that's currently commented out.
TestGreekUpper tests a separate code path in ICU4C that I haven't ported yet.

Will make a note to revisit this in the future.